### ISSUE TITLE
Failed expectation check on global mock object

### ISSUE THREAD

**Issue by wanghaiqiangk** [google/googletest#3410]
First please let me ask an unrelated question on the issue tracking system.
It seems that now only bug-labeled issues are allowed to post because of issue template, isn't it?
Since this post is more like to be a question, not a bug.

Back to the formal topic...
Maybe it's not a good practice to use global mock object, but no one prevents that to happen.
If I have more expectations than method invocations, it should have reported a failure such as saying "expect to be called once but never called", but actually it doesn't and ends up with a segmentation fault. This is a test program ->

```c++
#include <gmock/gmock.h>
#include <gtest/gtest.h>

using ::testing::_;
using ::testing::Return;
using ::testing::Mock;

class FooMock
{
public:
  MOCK_METHOD(bool, FooMethod, (int));
};

FooMock foomock; // global mock object                                                                                                  

TEST(FooTest, Counter)
{
  EXPECT_CALL(foomock, FooMethod(_))
    .WillRepeatedly(Return(true)); // a default expectation                                                                             

  EXPECT_CALL(foomock, FooMethod(1))
    .WillOnce(Return(false));

  EXPECT_CALL(foomock, FooMethod(2))
    .WillOnce(Return(true));

  EXPECT_CALL(foomock, FooMethod(3))
    .WillOnce(Return(false));

  EXPECT_EQ(foomock.FooMethod(1), false);
  EXPECT_EQ(foomock.FooMethod(2), true);
}

int main(int argc, char *argv[])
{
  testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}
```

Then compile and run. It gives me such output ->

```
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from FooTest
[ RUN      ] FooTest.Counter
[       OK ] FooTest.Counter (1 ms)
[----------] 1 test from FooTest (1 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (1 ms total)
[  PASSED  ] 1 test.
Segmentation fault (core dumped)
```

According to documentation, the expectation check is performed when the mock object is destroyed.  But since the mock object is in global scope, it's supposed to be checked when the program is to end its life. However, the extra expect_call along with resources allocated have been out of scope so that resulting into segmentation fault. Is it correct? I used gdb and did find the problem happens on the exit of program.

Then, with the effort of search, I think this can be solved by advancing the check stage via `Mock::VerifyAndClearExpectations()`. But the documentation indicates that

> Do not set new expectations after verifying and clearing a mock after its use. Setting expectations after code that exercises the mock has undefined behavior.

The statement says that it's undefined to set new expectations after `Mock::VerifyAndClearExpectations()` call. Maybe this is caused from some technical reasons, but this is not conform to common senses, at least to me. And I tried to add additional expectations to the same global mock object and it ran well.

```c++
#include <gmock/gmock.h>
#include <gtest/gtest.h>

using ::testing::_;
using ::testing::Return;
using ::testing::InvokeWithoutArgs;
using ::testing::Mock;

class FooMock
{
public:
  MOCK_METHOD(bool, FooMethod, (int));
};

FooMock foomock;                // global mock object                                                                                   

TEST(FooTest, Counter)
{
  EXPECT_CALL(foomock, FooMethod(_))
    .WillRepeatedly(Return(true)); // default expectation                                                                               

  EXPECT_CALL(foomock, FooMethod(1))
    .WillOnce(Return(false));

  EXPECT_CALL(foomock, FooMethod(2))
    .WillOnce(Return(true));

  EXPECT_CALL(foomock, FooMethod(3))
    .WillOnce(Return(false));

  EXPECT_EQ(foomock.FooMethod(1), false);
  EXPECT_EQ(foomock.FooMethod(2), true);

  Mock::VerifyAndClearExpectations(&foomock);
}

TEST(FooTest, Other)
{
  EXPECT_CALL(foomock, FooMethod(4))
    .WillOnce(Return(true));

  EXPECT_EQ(foomock.FooMethod(4), true);
}

int main(int argc, char *argv[])
{
  testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}
```

```
[==========] Running 2 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 2 tests from FooTest
[ RUN      ] FooTest.Counter
/home/wang/workspace/gtest-learn/test.cpp:28: Failure
Actual function call count doesn't match EXPECT_CALL(foomock, FooMethod(3))...
         Expected: to be called once
           Actual: never called - unsatisfied and active
[  FAILED  ] FooTest.Counter (0 ms)
[ RUN      ] FooTest.Other
[       OK ] FooTest.Other (0 ms)
[----------] 2 tests from FooTest (0 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test suite ran. (0 ms total)
[  PASSED  ] 1 test.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] FooTest.Counter

 1 FAILED TEST
```

So my question is
- Should I use `Mock::VerifyAndClearExpectations()`?
- Should I set more expectations after above call?
- Should I use global mock object? Maybe a test fixture is better right?

Additional information:
- googletest/mock version 1.10.0 release
- OS ubuntu 18.04.4
- Kernel 4.15.0-91-generic x86_64
- GCC Ubuntu 7.5.0-3ubuntu1~18.04
- cmake version 3.10.2

```CMake
cmake_minimum_required(VERSION 3.0)
project(gtest-learn)

set(CMAKE_CXX_FLAGS -g)

include_directories(/path/to/gtest/include)
link_directories(/path/to/gtest/lib)
set(GTEST_LIB -lgtest -lgmock -lpthread)

add_executable(runtest
    test.cpp)
target_link_libraries(runtest
    ${GTEST_LIB})
```

**Comment by derekmauro** [comment]
Is there a reason you are trying to use a global mock object? If not, use a test fixture instead.

**Comment by wanghaiqiangk** [comment]
> Is there a reason you are trying to use a global mock object? If not, use a test fixture instead.

Because I'm trying to mock a pure C function by providing a fake implementation in which the global mock object is called. To illustrate that,

```c
global_mock g_mock;

void pure_c_function(void)
{
    g_mock.pure_c_function();
}
```

And after replacing the link to real impl with this fake impl, I can test a C program.

Googlemock framework is designed to work with C++, so that I do not think it's the best way to apply in C. But since my project has chosen it as the base unit test framework, I have to think about a method to test some C functions. That's why I use global mock object, to provide fake implementations.

**Comment by asoffer** [comment]
You don't need to use a mock object to inject behavior, you only need a customization point that can call a function indirectly. In a lot of C++ code, we do this with virtual functions, but in C you can do this by passing around function pointers. I'd suggest, wherever `pure_c_function` is called, passing in a function pointer in place of that. You can pass in `pure_c_function` in real code, and `fake_pure_c_function` in test code, defining it to be whatever you want.

For example:
```
// This is taking the place of `pure_c_function`
int roll_die() { return rand(0) % 6 + 1; }

// Function we're testing
int roll_two_dice(int(*dice_roller)()) {
  return dice_roller() + dice_roller();
}

// In production code:
int result = roll_two_dice(roll_die);

// In tests
int roll_fake_die() { return 4; }

TEST(DiceRoller, Twice) {
  EXPECT_EQ(roll_two_dice(roll_fake_die), 8);
}
```

As for why this is undefined, I'm not entirely sure, but I suspect it would be dangerous... if expectations had failed, we cannot be sure the mock is in any sort of a valid state, and so resetting it may not be possible.

Hope this helps!

**Comment by hyukmyeong** [comment]
Hello,

I'm not sure about `Mock::VerifyAndClearExpectations()`, but an example of `Mock::VerifyAndClear()` can be found below
https://github.com/google/googletest/blob/master/googlemock/test/gmock-spec-builders_test.cc#L2572

And I think it would be better to use a global pointer and allocate/free in each test funciton!

**Comment by wanghaiqiangk** [comment]
@asoffer Thanks for the illustration of the injection tech. However, this idea probably is not much practical because we seldom define a function prototype with function pointer as parameter, for testing purpose. At least this is the case for the projects yet that I have been working on.

---

@hyukmyeong Thanks for providing the example. Regarding the aspect of behavior, the difference between `Mock::VerifyAndClearExpectations()` and `Mock::VerifyAndClear()` is whether to clear the *default* expectations. So this is minimal and I highly suspect that what can apply to the latter can also be adapted to the former.

---

Basically, the problem is derived from mocking a C routine and using a cpp mock framework like googlemock is probably not the best solution to this scene. How do you folks test legacy codes especially written in C?

**Comment by hyukmyeong** [comment]
I personally implement it like below
and I'm also curious how other people implement it

```cpp
#include <iostream>
#include <memory>
#include <unistd.h>
#include "gtest/gtest.h"
#include "gmock/gmock.h"

class SUT {
public:
  int foo(useconds_t param) { return usleep(param); }
};

class Mock {
public:
  static std::unique_ptr<Mock> mock;
  MOCK_METHOD1(usleep, int(useconds_t));
};
std::unique_ptr<Mock> Mock::mock(nullptr);

int usleep(useconds_t useconds) {
  std::cout << "I don't sleep" << std::endl;
  return Mock::mock->usleep(useconds);
}

class TestSUT : public ::testing::Test {
protected:
  void SetUp() {
    Mock::mock = std::unique_ptr<Mock>(new Mock());
  }
  void TearDown() {
    Mock::mock = nullptr;
  }
};

TEST_F(TestSUT, Foo) {
  // Arrange
  SUT sut;

  // Expect
  EXPECT_CALL(*Mock::mock, usleep(100))
    .WillOnce(::testing::Return(0));

  // Act
  sut.foo(100);
}
```
